File: api/src/adapters/registry.ts
/**
 * LLM-Registry Adapter
 *
 * Thin adapter for consuming model metadata and pipeline descriptors
 * from LLM-Registry.
 */

import { logger } from '../logger';
import {
  AdapterHealth,
  EcosystemAdapter,
  healthyAdapter,
  unhealthyAdapter,
} from './types';

export interface RegistryConfig {
  endpoint: string;
  apiKey?: string;
  timeoutMs?: number;
}

/**
 * Model metadata
 */
export interface ModelMetadata {
  modelId: string;
  name: string;
  version: string;
  provider: string;
  modelType: ModelType;
  capabilities: ModelCapability[];
  contextWindow: number;
  pricing: ModelPricing;
  performance: ModelPerformance;
  status: ModelStatus;
  registeredAt: Date;
  lastUpdated: Date;
  tags: Record<string, string>;
}

export type ModelType =
  | 'text_generation'
  | 'text_embedding'
  | 'image_generation'
  | 'image_analysis'
  | 'audio_transcription'
  | 'audio_generation'
  | 'multi_modal'
  | 'code_generation'
  | 'fine_tuned';

export type ModelCapability =
  | 'chat'
  | 'completion'
  | 'embedding'
  | 'function_calling'
  | 'vision'
  | 'audio'
  | 'streaming'
  | 'batch_processing'
  | 'fine_tuning';

export interface ModelPricing {
  currency: string;
  inputCostPer1kTokens: number;
  outputCostPer1kTokens: number;
  imageCostPerUnit?: number;
  audioCostPerMinute?: number;
}

export interface ModelPerformance {
  avgLatencyMs: number;
  p95LatencyMs: number;
  p99LatencyMs: number;
  tokensPerSecond: number;
  availability: number;
}

export type ModelStatus = 'active' | 'deprecated' | 'preview' | 'maintenance' | 'retired';

/**
 * Pipeline descriptor
 */
export interface PipelineDescriptor {
  pipelineId: string;
  name: string;
  version: string;
  description: string;
  stages: PipelineStage[];
  inputSchema: Record<string, unknown>;
  outputSchema: Record<string, unknown>;
  createdAt: Date;
  lastUpdated: Date;
  owner: string;
  status: PipelineStatus;
  metrics: PipelineMetrics;
}

export interface PipelineStage {
  stageId: string;
  stageName: string;
  stageType: StageType;
  modelId?: string;
  config: Record<string, unknown>;
  timeoutMs: number;
  retryPolicy: RetryPolicy;
}

export type StageType =
  | 'model_inference'
  | 'preprocessing'
  | 'postprocessing'
  | 'embedding'
  | 'retrieval'
  | 'routing'
  | 'aggregation'
  | 'transform'
  | 'cache'
  | 'rate_limit';

export interface RetryPolicy {
  maxAttempts: number;
  initialDelayMs: number;
  maxDelayMs: number;
  backoffMultiplier: number;
}

export type PipelineStatus = 'active' | 'paused' | 'draft' | 'archived';

export interface PipelineMetrics {
  totalInvocations: number;
  successRate: number;
  avgLatencyMs: number;
  avgCostPerInvocation: number;
}

/**
 * Provider info
 */
export interface ProviderInfo {
  providerId: string;
  name: string;
  status: 'operational' | 'degraded' | 'outage';
  apiVersion: string;
  models: string[];
  rateLimits: RateLimits;
  health: ProviderHealth;
}

export interface RateLimits {
  requestsPerMinute: number;
  tokensPerMinute: number;
  tokensPerDay?: number;
}

export interface ProviderHealth {
  availability: number;
  avgLatencyMs: number;
  errorRate: number;
  lastChecked: Date;
}

/**
 * Query parameters for models
 */
export interface ModelQuery {
  providers?: string[];
  modelTypes?: ModelType[];
  capabilities?: ModelCapability[];
  status?: ModelStatus;
  minContextWindow?: number;
  tags?: Record<string, string>;
}

/**
 * Query parameters for pipelines
 */
export interface PipelineQuery {
  owner?: string;
  status?: PipelineStatus;
  modelIds?: string[];
  createdAfter?: Date;
}

/**
 * LLM-Registry adapter for consuming registry data
 */
export class RegistryAdapter implements EcosystemAdapter {
  private config: RegistryConfig;
  private connected: boolean = false;

  constructor(config?: Partial<RegistryConfig>) {
    this.config = {
      endpoint: process.env.REGISTRY_ENDPOINT || 'http://localhost:8084',
      apiKey: process.env.REGISTRY_API_KEY,
      timeoutMs: parseInt(process.env.REGISTRY_TIMEOUT_MS || '30000', 10),
      ...config,
    };
  }

  async connect(): Promise<void> {
    logger.info({ endpoint: this.config.endpoint }, 'Connecting to LLM-Registry');
    this.connected = true;
    logger.info('Successfully connected to LLM-Registry');
  }

  async healthCheck(): Promise<AdapterHealth> {
    const start = Date.now();

    if (!this.connected) {
      return unhealthyAdapter('registry', 'Not connected');
    }

    const latencyMs = Date.now() - start;
    return healthyAdapter('registry', latencyMs);
  }

  async disconnect(): Promise<void> {
    logger.info('Disconnecting from LLM-Registry');
    this.connected = false;
  }

  /**
   * Fetch model by ID
   */
  async fetchModel(modelId: string): Promise<ModelMetadata> {
    if (!this.connected) {
      throw new Error('Registry adapter not connected');
    }

    logger.debug({ modelId }, 'Fetching model from Registry');

    // Placeholder implementation
    return {
      modelId,
      name: modelId,
      version: '1.0.0',
      provider: 'unknown',
      modelType: 'text_generation',
      capabilities: [],
      contextWindow: 0,
      pricing: {
        currency: 'USD',
        inputCostPer1kTokens: 0,
        outputCostPer1kTokens: 0,
      },
      performance: {
        avgLatencyMs: 0,
        p95LatencyMs: 0,
        p99LatencyMs: 0,
        tokensPerSecond: 0,
        availability: 0,
      },
      status: 'active',
      registeredAt: new Date(),
      lastUpdated: new Date(),
      tags: {},
    };
  }

  /**
   * List models
   */
  async listModels(query?: ModelQuery): Promise<ModelMetadata[]> {
    if (!this.connected) {
      throw new Error('Registry adapter not connected');
    }

    logger.debug({ query }, 'Listing models from Registry');

    // Placeholder implementation
    return [];
  }

  /**
   * Fetch pipeline by ID
   */
  async fetchPipeline(pipelineId: string): Promise<PipelineDescriptor> {
    if (!this.connected) {
      throw new Error('Registry adapter not connected');
    }

    logger.debug({ pipelineId }, 'Fetching pipeline from Registry');

    // Placeholder implementation
    return {
      pipelineId,
      name: pipelineId,
      version: '1.0.0',
      description: '',
      stages: [],
      inputSchema: {},
      outputSchema: {},
      createdAt: new Date(),
      lastUpdated: new Date(),
      owner: 'unknown',
      status: 'active',
      metrics: {
        totalInvocations: 0,
        successRate: 0,
        avgLatencyMs: 0,
        avgCostPerInvocation: 0,
      },
    };
  }

  /**
   * List pipelines
   */
  async listPipelines(query?: PipelineQuery): Promise<PipelineDescriptor[]> {
    if (!this.connected) {
      throw new Error('Registry adapter not connected');
    }

    logger.debug({ query }, 'Listing pipelines from Registry');

    // Placeholder implementation
    return [];
  }

  /**
   * Fetch provider info
   */
  async fetchProvider(providerId: string): Promise<ProviderInfo> {
    if (!this.connected) {
      throw new Error('Registry adapter not connected');
    }

    logger.debug({ providerId }, 'Fetching provider from Registry');

    // Placeholder implementation
    return {
      providerId,
      name: providerId,
      status: 'operational',
      apiVersion: '1.0',
      models: [],
      rateLimits: {
        requestsPerMinute: 0,
        tokensPerMinute: 0,
      },
      health: {
        availability: 0,
        avgLatencyMs: 0,
        errorRate: 0,
        lastChecked: new Date(),
      },
    };
  }

  /**
   * List providers
   */
  async listProviders(): Promise<ProviderInfo[]> {
    if (!this.connected) {
      throw new Error('Registry adapter not connected');
    }

    logger.debug('Listing providers from Registry');

    // Placeholder implementation
    return [];
  }
}

File: src/adapters/registry.rs
//! LLM-Registry Adapter
//!
//! Thin adapter for consuming model metadata and pipeline descriptors
//! from LLM-Registry.
//!
//! This adapter provides read-only access to registry data for analytics
//! purposes without modifying any upstream logic.

use super::{AdapterHealth, EcosystemAdapter};
use anyhow::Result;
use async_trait::async_trait;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::time::Instant;
use tracing::{debug, info, instrument};

/// Configuration for Registry adapter
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegistryConfig {
    pub endpoint: String,
    pub api_key: Option<String>,
    pub timeout_secs: u64,
}

impl RegistryConfig {
    pub fn from_env() -> Result<Self> {
        Ok(Self {
            endpoint: std::env::var("REGISTRY_ENDPOINT")
                .unwrap_or_else(|_| "http://localhost:8084".to_string()),
            api_key: std::env::var("REGISTRY_API_KEY").ok(),
            timeout_secs: std::env::var("REGISTRY_TIMEOUT_SECS")
                .unwrap_or_else(|_| "30".to_string())
                .parse()
                .unwrap_or(30),
        })
    }
}

/// Model metadata from registry
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelMetadata {
    pub model_id: String,
    pub name: String,
    pub version: String,
    pub provider: String,
    pub model_type: ModelType,
    pub capabilities: Vec<ModelCapability>,
    pub context_window: u64,
    pub pricing: ModelPricing,
    pub performance: ModelPerformance,
    pub status: ModelStatus,
    pub registered_at: DateTime<Utc>,
    pub last_updated: DateTime<Utc>,
    pub tags: HashMap<String, String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ModelType {
    TextGeneration,
    TextEmbedding,
    ImageGeneration,
    ImageAnalysis,
    AudioTranscription,
    AudioGeneration,
    MultiModal,
    CodeGeneration,
    FineTuned,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ModelCapability {
    Chat,
    Completion,
    Embedding,
    FunctionCalling,
    Vision,
    Audio,
    Streaming,
    BatchProcessing,
    FineTuning,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelPricing {
    pub currency: String,
    pub input_cost_per_1k_tokens: f64,
    pub output_cost_per_1k_tokens: f64,
    pub image_cost_per_unit: Option<f64>,
    pub audio_cost_per_minute: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelPerformance {
    pub avg_latency_ms: f64,
    pub p95_latency_ms: f64,
    pub p99_latency_ms: f64,
    pub tokens_per_second: f64,
    pub availability: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ModelStatus {
    Active,
    Deprecated,
    Preview,
    Maintenance,
    Retired,
}

/// Pipeline descriptor
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PipelineDescriptor {
    pub pipeline_id: String,
    pub name: String,
    pub version: String,
    pub description: String,
    pub stages: Vec<PipelineStage>,
    pub input_schema: serde_json::Value,
    pub output_schema: serde_json::Value,
    pub created_at: DateTime<Utc>,
    pub last_updated: DateTime<Utc>,
    pub owner: String,
    pub status: PipelineStatus,
    pub metrics: PipelineMetrics,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PipelineStage {
    pub stage_id: String,
    pub stage_name: String,
    pub stage_type: StageType,
    pub model_id: Option<String>,
    pub config: HashMap<String, serde_json::Value>,
    pub timeout_ms: u64,
    pub retry_policy: RetryPolicy,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum StageType {
    ModelInference,
    Preprocessing,
    Postprocessing,
    Embedding,
    Retrieval,
    Routing,
    Aggregation,
    Transform,
    Cache,
    RateLimit,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RetryPolicy {
    pub max_attempts: u32,
    pub initial_delay_ms: u64,
    pub max_delay_ms: u64,
    pub backoff_multiplier: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum PipelineStatus {
    Active,
    Paused,
    Draft,
    Archived,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PipelineMetrics {
    pub total_invocations: u64,
    pub success_rate: f64,
    pub avg_latency_ms: f64,
    pub avg_cost_per_invocation: f64,
}

/// Provider information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderInfo {
    pub provider_id: String,
    pub name: String,
    pub status: ProviderStatus,
    pub api_version: String,
    pub models: Vec<String>,
    pub rate_limits: RateLimits,
    pub health: ProviderHealth,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ProviderStatus {
    Operational,
    Degraded,
    Outage,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RateLimits {
    pub requests_per_minute: u64,
    pub tokens_per_minute: u64,
    pub tokens_per_day: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderHealth {
    pub availability: f64,
    pub avg_latency_ms: f64,
    pub error_rate: f64,
    pub last_checked: DateTime<Utc>,
}

/// Query parameters for models
#[derive(Debug, Clone, Default)]
pub struct ModelQuery {
    pub providers: Option<Vec<String>>,
    pub model_types: Option<Vec<ModelType>>,
    pub capabilities: Option<Vec<ModelCapability>>,
    pub status: Option<ModelStatus>,
    pub min_context_window: Option<u64>,
    pub tags: Option<HashMap<String, String>>,
}

/// Query parameters for pipelines
#[derive(Debug, Clone, Default)]
pub struct PipelineQuery {
    pub owner: Option<String>,
    pub status: Option<PipelineStatus>,
    pub model_ids: Option<Vec<String>>,
    pub created_after: Option<DateTime<Utc>>,
}

/// LLM-Registry adapter for consuming registry data
pub struct RegistryAdapter {
    config: RegistryConfig,
    connected: AtomicBool,
}

impl RegistryAdapter {
    pub fn new(config: RegistryConfig) -> Self {
        Self {
            config,
            connected: AtomicBool::new(false),
        }
    }

    /// Fetch model metadata by ID
    #[instrument(skip(self))]
    pub async fn fetch_model(&self, model_id: &str) -> Result<ModelMetadata> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Registry adapter not connected");
        }

        debug!(model_id = %model_id, "Fetching model metadata from Registry");

        // Placeholder implementation
        Ok(ModelMetadata {
            model_id: model_id.to_string(),
            name: model_id.to_string(),
            version: "1.0.0".to_string(),
            provider: "unknown".to_string(),
            model_type: ModelType::TextGeneration,
            capabilities: Vec::new(),
            context_window: 0,
            pricing: ModelPricing {
                currency: "USD".to_string(),
                input_cost_per_1k_tokens: 0.0,
                output_cost_per_1k_tokens: 0.0,
                image_cost_per_unit: None,
                audio_cost_per_minute: None,
            },
            performance: ModelPerformance {
                avg_latency_ms: 0.0,
                p95_latency_ms: 0.0,
                p99_latency_ms: 0.0,
                tokens_per_second: 0.0,
                availability: 0.0,
            },
            status: ModelStatus::Active,
            registered_at: Utc::now(),
            last_updated: Utc::now(),
            tags: HashMap::new(),
        })
    }

    /// List models matching query
    #[instrument(skip(self))]
    pub async fn list_models(&self, query: ModelQuery) -> Result<Vec<ModelMetadata>> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Registry adapter not connected");
        }

        debug!("Listing models from Registry");

        // Placeholder implementation
        Ok(Vec::new())
    }

    /// Fetch pipeline descriptor by ID
    #[instrument(skip(self))]
    pub async fn fetch_pipeline(&self, pipeline_id: &str) -> Result<PipelineDescriptor> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Registry adapter not connected");
        }

        debug!(pipeline_id = %pipeline_id, "Fetching pipeline descriptor from Registry");

        // Placeholder implementation
        Ok(PipelineDescriptor {
            pipeline_id: pipeline_id.to_string(),
            name: pipeline_id.to_string(),
            version: "1.0.0".to_string(),
            description: String::new(),
            stages: Vec::new(),
            input_schema: serde_json::json!({}),
            output_schema: serde_json::json!({}),
            created_at: Utc::now(),
            last_updated: Utc::now(),
            owner: "unknown".to_string(),
            status: PipelineStatus::Active,
            metrics: PipelineMetrics {
                total_invocations: 0,
                success_rate: 0.0,
                avg_latency_ms: 0.0,
                avg_cost_per_invocation: 0.0,
            },
        })
    }

    /// List pipelines matching query
    #[instrument(skip(self))]
    pub async fn list_pipelines(&self, query: PipelineQuery) -> Result<Vec<PipelineDescriptor>> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Registry adapter not connected");
        }

        debug!("Listing pipelines from Registry");

        // Placeholder implementation
        Ok(Vec::new())
    }

    /// Get provider information
    #[instrument(skip(self))]
    pub async fn fetch_provider(&self, provider_id: &str) -> Result<ProviderInfo> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Registry adapter not connected");
        }

        debug!(provider_id = %provider_id, "Fetching provider info from Registry");

        // Placeholder implementation
        Ok(ProviderInfo {
            provider_id: provider_id.to_string(),
            name: provider_id.to_string(),
            status: ProviderStatus::Operational,
            api_version: "1.0".to_string(),
            models: Vec::new(),
            rate_limits: RateLimits {
                requests_per_minute: 0,
                tokens_per_minute: 0,
                tokens_per_day: None,
            },
            health: ProviderHealth {
                availability: 0.0,
                avg_latency_ms: 0.0,
                error_rate: 0.0,
                last_checked: Utc::now(),
            },
        })
    }

    /// List all providers
    #[instrument(skip(self))]
    pub async fn list_providers(&self) -> Result<Vec<ProviderInfo>> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Registry adapter not connected");
        }

        debug!("Listing providers from Registry");

        // Placeholder implementation
        Ok(Vec::new())
    }
}

#[async_trait]
impl EcosystemAdapter for RegistryAdapter {
    #[instrument(skip(self))]
    async fn connect(&self) -> Result<()> {
        info!(endpoint = %self.config.endpoint, "Connecting to LLM-Registry");

        self.connected.store(true, Ordering::Relaxed);

        info!("Successfully connected to LLM-Registry");
        Ok(())
    }

    async fn health_check(&self) -> Result<AdapterHealth> {
        let start = Instant::now();

        if !self.connected.load(Ordering::Relaxed) {
            return Ok(AdapterHealth::unhealthy("registry", "Not connected"));
        }

        let latency_ms = start.elapsed().as_millis() as u64;
        Ok(AdapterHealth::healthy("registry", latency_ms))
    }

    #[instrument(skip(self))]
    async fn disconnect(&self) -> Result<()> {
        info!("Disconnecting from LLM-Registry");
        self.connected.store(false, Ordering::Relaxed);
        Ok(())
    }
}
